From dfdd9bd420f4e9c9b96891e5ea072b1cf9ddb2d3 Mon Sep 17 00:00:00 2001
From: boby <glboby27@gmail.com>
Date: Tue, 11 May 2021 21:04:12 -0700
Subject: [PATCH 3/3] always-write-an-stss-box-in-CMAF-mode

---
 libavformat/movenc.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/libavformat/movenc.c b/libavformat/movenc.c
index c6acbcebfe..2f84b142ff 100644
--- a/libavformat/movenc.c
+++ b/libavformat/movenc.c
@@ -2623,6 +2623,10 @@ static int mov_write_stbl_tag(AVFormatContext *s, AVIOContext *pb, MOVMuxContext
          track->par->codec_tag == MKTAG('r','t','p',' ')) &&
         track->has_keyframes && track->has_keyframes < track->entry)
         mov_write_stss_tag(pb, track, MOV_SYNC_SAMPLE);
+    if (track->par->codec_type == AVMEDIA_TYPE_VIDEO &&
+        mov->flags & FF_MOV_FLAG_CMAF &&
+        !track->has_keyframes)
+        mov_write_stss_tag(pb, track, MOV_SYNC_SAMPLE);
     if (track->par->codec_type == AVMEDIA_TYPE_VIDEO && track->has_disposable)
         mov_write_sdtp_tag(pb, track);
     if (track->mode == MODE_MOV && track->flags & MOV_TRACK_STPS)
-- 
2.25.1

